### PR TITLE
Enable Pinocchio to run on anvil

### PR DESCRIPTION
**Pull request by DeaglanBartlett**
To allow Pinocchio to run on anvil, I have:

- Enabled the executable to be run with `mpirun`. This automatically checks the number of available processes available for a PBS, SLURM or local job.
- Updated the documentation to explain which modules are needed to compile Pinocchio and what must be included in the Makefile.

### CODE REVIEW

**Comment by DeaglanBartlett** [comment]
For reference, for lhid=3 of the quijote LH simulations, here is how the HMF and power spectrum compares between the truth and the Pinocchio runs. Ideally we would want to get to below $10^{13}$ in halo mass, so we may need to run higher than $N=384$. But $N=384$ looks reasonably good for above  $\sim 10^{13.4}$.

pinocchio_hmf_3
pinocchio_halo_pk_split_3
pinocchio_halo_pk_13 4_3

**Comment by DeaglanBartlett** [comment]
For reference, running the quijote-like simulations on 20 cores requires

| N   | Required memory per task (Gb)   | Total required memory (Gb)   |
|------------|------------|------------|
| 128 | 0.028 | 0.543457|
| 256 | 0.223 | 4.347656|
| 384 | 0.747 | 14.582016|
| 512 | 1.772| 34.618896|

In fact, using just 8 cores can be enough to keep the required memory per task below 2Gb when going up to N=384 , but for N=512, you need at least 18 cores to achieve this.

Although this is calculated on the fly in the Pinocchio run to check the resources are there, to check in advance, one needs to run
```
mpirun -np 1 $PINOCCHIO_DIR/src/memorytest parameter_file $NUMBER_OF_CORES
```
for the appropriate Pinocchio parameter_file

**Comment by DeaglanBartlett** [comment]
Running with $N=512$ gives good results to down to at least $M_{\rm h} = 10^{13}  h^{-1} {\rm M_{\odot}}$ (note all earlier plots missing the $h$ in the units, but this was there).

pinocchio_hmf_3 (2)
pinocchio_halo_pk_split_3 (2)
pinocchio_halo_pk_13 0_3 (1)

Note that for $N=512$ I used random ICs with the same cosmology as lhid=3, since I did not have access to the true white noise field at that resolution.

**Comment by DeaglanBartlett** [comment]
As an estimate, running the `memorytest` script, to get the memory requirements below 2Gb per core would require at least the following number of cores (nc) for box lengths of L and values of N (such that L/N is the same as the 1Gpc/h run above)

```
L = 2Gpc/h, N=1024, nc=144
L = 3Gpc/h, N=1536, nc=480
```

Note though that these are probably lower limits for the number of cores - we needed more than the estimate 20 above to do 1Gpc/h.

The high memory queues on anvil have 8Gb per core rather than 2 (but then are charged at four times the cost of the normal queue). If we used this queue, then we would need approximately

```
L = 1Gpc/h, N=512, nc=4
L = 2Gpc/h, N=1024, nc=35
L = 3Gpc/h, N=1536, nc=120
```

**NOTE:** The memory requirement does not monotonically decrease as one increases the value of nc, since the splitting of planes changes. It does look like, though, that we may be able to fit the 3Gpc/h run onto a single node if needed if using the high memory nodes. But maybe we are just happy using multiple nodes.

**Comment by DeaglanBartlett** [comment]
I have now tested running the abacus-like (2Gpc/h) simulations on anvil. This required some changes as the number of particles was too large to be handled correctly by the original post-processing routines. I ran with the quijote lhid=3 but random ics using 2 nodes with 256 tasks. 

The total run time was 8min26s, of which 7min51s was spent in "main".

The output directory ended up being of size 50G, but the important products (nbody.h5 and halos.h5) come to ~0.5G of this, and we can probably delete the rest. Note that I decided to save the nbody on a grid of size Ngrid=256 for this rather than using N=1024 as this was quite memory intensive, and I don't believe we will need this resolution.

Checking the hmf against Quijote and the Quijote-like pinocchio run gives good results:
pinocchio_abacus_quijote_hmf_3

and the power spectrum looks similar to the Quijote-like pinocchio run (I saved the latter with finer resolution, hence it can go to smaller k. The ICs are not matched, so the large scales slightly differ too):
pinocchio_abacus_quijote_pk_3

File: .gitignore [hunk 1]
```diff
@@ -139,3 +139,6 @@ timing_stats*
 sbi-logs
 matts_tests
 .vscode
+
+cmass/conf/nbody/pin_1gpch*.yaml
```
Comment (maho3) [inline .gitignore:143]: Is this still necessary, or only for testing?

File: cmass/conf/nbody/pinocchio_2gpch.yaml
Comment (maho3) [inline cmass/conf/nbody/pinocchio_2gpch.yaml:1]: Can these two yamls match the structure of the other nbody configs? Before merging

File: jobs/make_symlinks.py
Comment (maho3) [inline jobs/make_symlinks.py:1]: Is the purpose of this to make copies of a simulation hierarchy, with only certain desired files?

File: jobs/pinocchio_quijote_0.sh
Comment (maho3) [inline jobs/pinocchio_quijote_0.sh:1]: Maybe clean these 0, 1, 2, 3 up before merging

File: jobs/setup_pinocchio_yaml.py
Comment (maho3) [inline jobs/setup_pinocchio_yaml.py:1]: This is cool but may not be useful for most people, apart from doing the limit testing that we are. Perhaps in the final version, we move this to ltu-gobig?

**Review by maho3 — COMMENTED** [review]
I left several comments in this review, just following from a read-only overview of the code. I will next try running it and leave more comments in the PINNOCHIO implementation

**Comment by DeaglanBartlett** [comment]
For the cmass runs, we will want to use the following snapshots:
```
asave: [0.76721, 0.74710, 0.72699, 0.70688, 0.68677, 0.66666, 0.64655, 0.62644, 0.60633, 0.58622]
```

Running multiple snapshots makes the code slower, such that the total time for `run_density` is now approximately 16-17min. I have adapted the processing of the output snapshots to run using MPI so that each core receives a sub-set of the output redshifts. Ideally one could use all the cores to do the construction of the density field, but this was the simplest way of using multiple cores without rewriting the density field constructor.

This commit is not yet ready to merge as the lock files do not seem to stop the different processes writing to the same file simultaneously

**Comment by maho3** [comment]
I've now got it working on my account on anvil and it works well! Just requires a little cleanup before merging.

**Review by maho3 — APPROVED** [review]

File: jobs/pinocchio_quijote_0.sh
Comment (DeaglanBartlett) [reply jobs/pinocchio_quijote_0.sh:1]: I have removed these files and kept one version as `jobs/pinocchio_quijote.sh`

File: .gitignore (hunk 1 above)
Comment (DeaglanBartlett) [reply .gitignore]: This has now been removed

File: jobs/make_symlinks.py
Comment (DeaglanBartlett) [reply jobs/make_symlinks.py:1]: Yes, but it is not generally useful so I have removed it

File: jobs/setup_pinocchio_yaml.py
Comment (DeaglanBartlett) [reply jobs/setup_pinocchio_yaml.py:1]: I agree. I've removed it for now, but kept a copy elsewhere so we could potentially move it there at some point

File: cmass/conf/nbody/pinocchio_2gpch.yaml
Comment (DeaglanBartlett) [reply cmass/conf/nbody/pinocchio_2gpch.yaml:1]: Done

**Pull request merged**